### PR TITLE
Create initial donor form with submission endpoint #14

### PR DESCRIPTION
**Pull request by wenta**

### CODE REVIEW

File: app/Http/Controllers/AdoptionController.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        return view('pages/adoptionForm', ['animalId' => $animalId]);
+    }
+
+
+    public function submitForm(Request $request)
+    {
+        // Validate form data
+        $validatedData = $request->validate([
+            // Validation rules for animal fields
+            'animalName' => 'required|max:255',
```
Comment (Guferos) [inline app/Http/Controllers/AdoptionController.php:24]: Please remove `animalName` from the validation, it's not required in the post data.

File: routes/web.php [hunk 2]
```diff
@@ -19,6 +20,8 @@
 Route::get('/psy', [AnimalController::class, 'dogsList'])->name('dogs');
 Route::get('/koty', [AnimalController::class, 'catsList'])->name('cats');
 Route::get('/zwierzak/{id}', [AnimalController::class, 'details'])->name('animal.details');
+Route::get('/adoption/{animalId}', [AdoptionController::class, 'index'])->name('adoption');
```
Comment (Guferos) [inline routes/web.php:23]: The donor's form should be accessible from the animal details page. Therefore this route will not be needed.

File: routes/web.php [hunk 3]
```diff
@@ -19,6 +20,8 @@
...
+Route::post('/submit-donor', [AdoptionController::class, 'submitForm'])->name('submitForm');
```
Comment (Guferos) [inline routes/web.php:24]: Rather then having separate AdoptionController, it feels more logical to move the submission into the DonorController into the `store` method. Could you please move AdoptionController logic there?

File: resources/views/pages/adoptionForm.blade.php
Comment (Guferos) [inline resources/views/pages/adoptionForm.blade.php:1]: Please move the form into the animalDetails.blade page. We should be accessing form from there (same as in Ciapkowo).

File: resources/views/pages/adoptionForm.blade.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+<html lang="pl">
+<head>
+    <meta charset="UTF-8">
+    <title>Formularz Adopcji Zwierząt</title>
+</head>
+<body>
+    <form action="/submit-donor" method="POST">
+        @csrf <!-- CSRF Token -->
+
+        <input type="hidden" name="animal_id" value="{{ $animalId }}">
```
Comment (Guferos) [inline resources/views/pages/adoptionForm.blade.php:15]: Once moved to the animalDetails.blade file, `animal_id` will be assigned from the `$animal` object available there.

File: resources/views/pages/adoptionForm.blade.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+        <label for="animalName">Imię zwierzęcia:</label>
```
Comment (Guferos) [inline resources/views/pages/adoptionForm.blade.php:18]: Animal name is not needed.

File: tests/Unit/AdoptionFormTest.php
Comment (Guferos) [inline tests/Unit/AdoptionFormTest.php:1]: Well done for the tests. Finally someone is doing them! I need to add some myself.

File: app/Http/Controllers/AdoptionController.php
Comment (Guferos) [inline app/Http/Controllers/AdoptionController.php:1]: As commented in other places. Please move the `submitForm` logic into the `DonorController->store` method. AdoptionController will not be needed really, we should be accessing the form from the `animalDetails.blade.php` instead.

File: resources/sass/app.scss
Comment (Guferos) [inline resources/sass/app.scss:1]: Screenshot 2023-11-16 at 08 28 27

UI it's not so important at this stage, but just wanted to mention that there is something wrong with the email field (it's too short and looks different to other fields in the form). I tested on Safari 16.4

**Review by Guferos — CHANGES_REQUESTED** [review]
Generally all good! It seems like you picking the Laravel skills very quickly. Thanks for the tests, finally someone done them! (I am the most quilty here for not doing them). 

There is one thing that needs to be done, the donor form should be on the animal details page rather than on it's own page. Sorry I probably should be more clearer on the task's card about this. 

See my detailed comments to see what should  be changed before merge. Thanks.

**Comment by Guferos** [comment]
@wenta Please merge the dev branch before submitting fixes to the pull and run database migrations (ideally as fresh) + seeders. I just merged a significant change to the database structure that removes `animal_id` from `Donor` model and moves Animals-Donors relationship into a pivot as we need Many to many relationship rather than One to Many we had previously. This would affect your donor form submission implementation. Rather than setting `animal_id` on Donor, you would need to attach (https://laravel.com/docs/10.x/eloquent-relationships#updating-many-to-many-relationships) Animal when creating a Donor record. 

On top of that there are two more things that need to be done when creating a donor:
1) You would need to validate if submitted animal_id is correct by checking if there is Animal record with the submitted id
2) As we are going to run two queries here (save Donor and attach Animal), you need to run both queries wrapped in database transaction to ensure data integrity if one of them fails.

**Comment by Guferos** [comment]
@wenta Thanks for the pr fixes. Not sure if you have seen my latest comment in this PR? Please have a look above. This branch must be merged with the `dev` first and some additional changes are required as I had to modify the way Donors are linked to Animals. It is now a many to many relationship rather than One to Many. Could you please look into that whenever possible?

**Comment by wenta** [comment]
Yes I work on that ;) I will let you know when pr will be finished,  I push commits for safety.

**Comment by Guferos** [comment]
@wenta Ah ok that's great, thanks. I wasn't sure if you pushed for a review or not.

**Review by Guferos — APPROVED** [review]
All good, thanks. We would probably need to replace the form with some sort of success message once the submission was successfull, but it wasn't part of this ticket, so I'll create a new one for that.

**Pull request merged**